File: python/chapter2/2-3-4.py
# 入力
N = int(input())
a = list(map(int, input().split()))

# 累積和を求める
S = [0] * (N + 1)
for i in range(N):
    S[i + 1] = S[i] + a[i]

# 配列 dp を定義する
INF = 2 ** 60  # 十分大きい値
dp = [[INF] * (N + 1) for i in range(N + 1)]

# 初期条件
for i in range(N):
    dp[i][i + 1] = 0

# dp[i][j] の値を順に求めていく
for between in range(2, N + 1):
    for i in range(0, N):
        j = i + between  # 区間の右端
        if j > N:
            break
        
        # 区間 [i, j) の分割位置で場合分けする
        for k in range(i + 1, j):
            dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j] + S[j] - S[i])

# 答え
print(dp[0][N])

File: cpp/chapter2/2-3-4.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main() {
    // 入力
    int N;
    cin >> N;
    vector<long long> a(N);
    for(int i = 0; i < N; i++) {
        cin >> a[i];
    }

    // 累積和を求める
    vector<long long> S(N + 1);
    for(int i = 0; i < N; i++) {
        S[i + 1] = S[i] + a[i];
    }

    // 配列 dp を定義する
    const long long INF = 1LL << 60; // 十分大きい値
    vector<vector<long long>> dp(N + 1, vector<long long>(N + 1, INF));

    // 初期条件
    for(int i = 0; i < N; i++) {
        dp[i][i + 1] = 0;
    }

    // dp[i][j] の値を順に求めていく
    for(int between = 2; between <= N; between++) {
        for(int i = 0; i < N; i++) {
            int j = i + between; // 区間の右端
            if(j > N) {
                break;
            }

            // 区間 [i, j) の分割位置で場合分けする
            for(int k = i + 1; k < j; k++) {
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j] + S[j] - S[i]);
            }
        }
    }

    // 答え
    cout << dp[0][N] << endl;
    return 0;
}
